pkg/proc/native/proc_linux.go

```diff
@@ -230,10 +231,20 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			if err != nil {
 				if err == sys.ESRCH {
 					// thread died while we were adding it
+					delete(dbp.threads, int(cloned))
 					continue
 				}
 				return nil, err
 			}
+			dbp.haltMu.Lock()
```

Comment:
We discussed briefly on IRC about this, but to bring it to this review:

I think we should be more careful of how we use this lock. Also, I'm not sure the value of `dbp.halt` anymore. It used to make sense because of the various implementations of `Halt` (which has changed a bit over time) because we needed to have state to signify that the inferior receiving `SIGSTOP` was actually sent via the debugger in attempt to stop the thread.

As a result of various changes I think we have 2 implementations of essentially the same thing, and their usage internally tends to be confusing and misleading. From an API / client perspective the only thing that gets called is `RequestManualStop`. The API command `halt` actually just calls `RequestManualStop`. We do not need to have two separate methods with 2 separate pieces of state (`halt` & `manualStopRequested`). It seems `(*Process) Halt()` is only used internally throughout proc and in tests.

Additionally, the comment for `RequestManualStop` is wrong, and the implementation, at least on Linux, could be improved. The comment mentions that we send `SIGSTOP` to all threads, when in actuality we only send `SIGTRAP` to the group leader. This may not be the best approach, as we essentially detect that in `trapWait` and then force all other threads to stop (by calling `thread.halt()`, which sends `SIGSTOP` to the thread and then immediately waits, so even there the `halt` flag is unnecessary. This is also inconsistent because I feel all waits should be filtered through the same code path, because what if the result of that wait is not the signal that we sent to the thread, but other important process state (we do specifically wait on the thread ID, but if that's the group leader, we could get other info).

To expand on the above paragraph, I feel we should combine the two methods and have on definitive way of stopping a running process. Also, on Linux the implementation could be improved (on OSX it's more difficult because of how ptrace / mach works). Currently we are sending a signal to the group leader putting it into a `signal-delivery-stop`, throwing away that signal (never delivering to the inferior) and then stopping all other threads within `trapWait`. What we could actually do is send `SIGSTOP`, and actually deliver it to the inferior which would cause all threads being traced to stop in a `group-stop`. I believe we would have to explicitly continue the process with `SIGCONT` as well on the flip side, but the benefit is all threads stop immediately.

One last thing: digging back through the code during this review I found something I'm really not a fan of and wish to change. I do not like that `ManualStopRequested` has the side effect of clearing `(*Process).manualStopRequested`. Predicate methods should not have side effects.

I'd like to have any discussion around the above comments in this PR, however any resulting action to be taken can happen in follow up PR(s).

Reply:
> As a result of various changes I think we have 2 implementations of essentially the same thing, and their usage internally tends to be confusing and misleading

Agree with this, I think `proc/native.(*Process).Halt` should be combined with `proc/native.(*Process).setCurrentBreakpoints` and just be some private method of proc/native with an OS specific implementation. The call to Halt in the tests can probably be removed entirely, I don't think we ever get to it with the process running.

> This is also inconsistent because I feel all waits should be filtered through the same code path, because what if the result of that wait is not the signal that we sent to the thread, but other important process state (we do specifically wait on the thread ID, but if that's the group leader, we could get other info).

This patch is doing that by calling trapWait after sending SIGSTOP. There's only one other place where we wait, that I can think of, and is in singlestep but that's probably fine given that everything else is stopped and the thread is in single step mode.

> What we could actually do is send SIGSTOP, and actually deliver it to the inferior which would cause all threads being traced to stop in a group-stop. I believe we would have to explicitly continue the process with SIGCONT as well on the flip side, but the benefit is all threads stop immediately.

I've checked the gdb native backend recently and as far as I can tell they are doing the signal delivery stop thing too, I'm not feeling too confident straying off the beaten path here.

> I do not like that ManualStopRequested has the side effect of clearing (*Process).manualStopRequested. Predicate methods should not have side effects.

Do you think CheckAndClearManualStopRequested would be a better name? Suggestions?

Reply:
> Do you think CheckAndClearManualStopRequested would be a better name? 

I'm fine with that, I don't verbosity for clarity.

> Agree with this, I think proc/native.(*Process).Halt should be combined with proc/native.(*Process).setCurrentBreakpoints and just be some private method of proc/native with an OS specific implementation. The call to Halt in the tests can probably be removed entirely, I don't think we ever get to it with the process running.

I think instead of combining we should just merge to a single implementation and call that where needed.

-------

I think overall we can merge this after the function rename, and then send a follow up PR and address the other concerns outlined about. What do you think @aarzilli?

Reply:
> I think instead of combining we should just merge to a single implementation and call that where needed.

what do you mean?

Reply:
> what do you mean?

Sorry, I wasn't very clear with that. What I meant was: it seemed you were suggesting moving the implementation of `halt` into `(*Process).setCurrentBreakpoints`, but to me that's only hiding the issue. We should have only one code path that knows how to stop a running process consistently, and simply call that routine where needed. Apologies if I misunderstood you.